text/0011-gem-checksum-verification.md

```diff
@@ -0,0 +1,217 @@
+This would make dependabot the defacto source of the checksums in the Gemfile instead of rubygems.org or .gem files.
+This might lead users to disable frozen bundles, disable dependabot, or disable the checksums feature, any of which could be considered major drawbacks.
+
+This can be addressed by making clear messaging about how to fix this problem (checkout the branch and run bundle install)
+Additionally, we should work with the dependabot team to provide the information necessary to add checksums to the lockfile.
+
+### Older versions of Bundler
+
+Old versions of Bundler should ignore the CHECKSUMS section.
+
+# Rationale and Alternatives
+
+Rubygems.org already stores SHA256 checksums for gems and returns them in the compact index response.
+All the information is already present for checksum verification on the client side.
+
+Verifying .gem files at install time offers nearly complete protection against hacked, altered and corrupted gems.
+Bundler already ensures that only the bundled gems are available to the app.
+This feature adds the assurance that only verified gems were installed with the bundle.
+
+One alternative to this solution is including (vendoring) the bundled gems in the repository.
+This effectively has the same result, since the gems that will be installed during the production deploy will be verifiably the same gems that were used during CI and development.
+The downside of this vendored approach is the increase in the repository size.
+The upside is installation that doesn't depend on a remote source.
+Checksums allow for a similar level of confidence without the larger repository size of vendoring gems.
+It can be enabled by default so that more users will benefit.
+
+# Unresolved questions
+
+### What are the unexpected errors that may happen the first time developers interact with this feature?
+
+In particular, the first deploy for users with a frozen bundle in CI and/or production may produce errors that might not have obvious solutions to someone unfamiliar with the feature.
+
+### How do we handle confusion about the authority of checksums written to the Gemfile.lock
+
+The source of checksums in the Gemfile.lock becomes a matter of trust once it's written.
+Did the checksum come from the API or was it calculated from a .gem file on a developers computer.
+If a checksum error is resolved by one developer in a way that saves an incorrect checksum, how should people know when to approve these changes or not.
+It may not be common practice for most teams to look at the Gemfile.lock when reviewing code.
+Gemfile.lock changes can be hidden in pull request reviews.
+Without a process for checking that the checksums are trustworthy, it's left to every development team to decide on a process.
+
+One solution would be a bundle command that could be run in CI every time the gems are installed that verifies the authenticity of checksums in the Gemfile.lock.
```

Comment:
Technically `bundle install` in frozen mode in CI would catch it, since there won't be a local gem cache?

Reply:
I meant a command for syncing checksums from the remote to ensure you don't have an erroneous checksum generated from a broken gem.

Reply:
mmmm, so a command that fixes the issue rather than detecting it, right? Maybe `bundle lock` without a local gem cache including that broken gem should do that, either by default or under an optional flag?

Reply:
Now that I think about it, I think it should sync and compare everything. This is the chance to print that "the gem you have installed generated a checksum that is different than what the server is saying."

Bundle lock should just pull all checksums every time then? Lock previously doesn't hit the network unless a resolve is necessary. This would make it always hit the network unless `--no-checksums` or `--local` is set.

Reply:
Ah, my explicit concern is this scenario:

1. Malicious developer sets checksum and upload gem.
2. Checksum would be invalid if compared with server, but doesn't happen during install.
3. Malicious gem is successfully installed.

If we had a process that could verify your checksums against the server without altering your bundle, it could run in CI as part of tests and help protect against malicious or bad checksums in lockfiles.

Reply:
I think it's fine for `bundle lock` to hit the network (or local cache) to look for checksums.

I'm not sure I understand your concern, can you elaborate this 3 steps? What does "sets checksum" mean? Manually edit a lockfile changing the checksum of a gem? What does "upload gem" mean in this context? Are you implying a situation where a malicious actor has both control of your lockfile, and your remote server? That sounds wild and I don't see how anything can be done in that situation 🤣

Overall, I think `bundle install --force` in frozen mode should be a strict way to check remote checksums against lockfile checksums, since that bypasses the local cache and should not allow lockfile changes.

Reply:
hehe, maybe my imagination is running a bit wild. With that access you can do whatever you want.

The presence of bundle install --force should provide options for people wanting to use this feature like that.

The more we talk about it, the more clear it is that we can't anticipate everything.

Reply:
> With that access you can do whatever you want.

Exactly, we shouldn't care about it since at that point there's no much we can do.